File: leetcode/Solution104.py
from collections import deque
from typing import Optional

from TreeNode import TreeNode

class Solution104:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
    #   DFS
       
        # if not root:
        #     return 0
        
        # return 1 + max(self.maxDepth(root.left), self.maxDepth(root.right))

        
        
        
    #   BFS
        
        # level = 0
        # q = deque([root])

        # while q:
        #     for i in range(len(q)):
        #         node = q.popleft()

        #         if node.left:
        #             q.append(node.left)
        #         if node.right:
        #             q.append(node.right)
            
        #     level += 1

        # return level

        
        # DFS Iterative

        stack = [[root,1]]
        result = 1
        while stack:
            node, depth = stack.pop()

            if node:
                result = max(result, depth)
                stack.append([node.left, depth + 1])
                stack.append([node.right, depth + 1])

        return result


s = Solution104()
s.maxDepth([3,9,20,None,None,15,7])

File: leetcode/Solution104.java
package leetcode;

import java.util.LinkedList;
import java.util.Queue;

class Solution104 {
    public int maxDepth(TreeNode root) {
        // DFS recursive

        // if (root == null) {
        // return 0;
        // }

        // return 1 + Math.max(maxDepth(root.left), maxDepth(root.right));

        // BFS

        Queue<TreeNode> queue = new LinkedList<TreeNode>();
        queue.add(root);
        int d = 0;
        while (!queue.isEmpty()) {
            int s = queue.size();
            for (int i = 0; i < s; i++) {
                TreeNode node = queue.poll();

                if (node.left != null) {
                    queue.add(node.left);
                }
                if (node.right != null) {
                    queue.add(node.right);
                }
            }
            d++;
        }

        return d;

    }
}